Webhook Improvements - Templateable Payloads

JacobsonMT opened the pull request:
## Overview
This PR introduces enhancements to the webhook notifier, focusing on three main areas:
1. Customizable webhook payloads using templates
2. Support for custom HTTP headers
3. New template functions to provide support for creating well-formatted, valid JSON in payloads.

## Key Changes

### 1. Custom Payload Templates
- Added support for custom JSON payload templates in webhook notifications
- Users can now define their own payload structure using templates
- Includes a new `CustomPayload` field with:
  - `Template` field for defining the payload structure
  - `Vars` field for providing additional template variables accessible in context via `.Vars.<variable_name>`
- Provides an example payload template which is equivalent to the default payload format for demonstration and regression testing

### 2. Custom HTTP Headers
- Added support for custom HTTP headers in webhook requests, example usage would be to set the `Content-Type` header for your custom payload.
- Implemented security measures to prevent misuse:
  - Restricted headers list to prevent security issues
  - Headers like `Authorization`, `Cookie`, `Host` are protected

### 3. New Template Functions
- Introduced new template functions inspired by [gomplate](https://docs.gomplate.ca/):
  - `coll` namespace for collection operations (`Dict`, `Slice`, `Append`)
  - `data` namespace for JSON operations (`JSON`, `ToJSON`, `ToJSONPretty`)
  - `tmpl` namespace for template operations compatible with pipelines (`Inline`, `Exec`)
  - `time` namespace for time operations (`Now`)
- These functions enable more complex and flexible payload formatting, especially when creating JSON.
- Currently gomplate does not allow modularized use of select functions or namespaces, so importing it directly would require a significant increase in dependencies. In addition, certain necessary functions have extended functionality in gomplate that makes them risky to use as-is such as the `eJSON` decryption capability of `data.JSON` which reads from environment variables. That being said, none of this is insurmountable, and the plan should be to eventually incorporate the gomplate library directly and more broadly.

### 4. Related Improvements
- Added new fields to `ExtendedData` and `ExtendedAlert` to make sure custom payloads have access to all of the information the default webhook payload has:
  - `ExtendedAlert.OrgId`
  - `ExtendedData.GroupKey`
  - `ExtendedData.TruncatedAlerts`

## Related Requests

Issues for supporting new notifiers in Grafana are fairly common and are generally reasonable requests. However, they can come with significant burden of maintenance. Some examples:
- [Telegram Local Bot API support](https://github.com/grafana/grafana/issues/81534)
- [Alerting: Extend Pagerduty notification for Openduty](https://github.com/grafana/grafana/issues/6504)
- [Feature request: New contact point type: Amazon EventBridge](https://github.com/grafana/grafana/issues/82645)
- [FR: Alerting Contact Point Matrix](https://github.com/grafana/grafana/issues/82764)
- [Add web push notifications to Grafana Alerting](https://github.com/grafana/grafana/issues/10306)

Requests for related features are some of the most popular in Alerting:
- [[Feature request] Custom json passed as a webhook alert](https://github.com/grafana/grafana/issues/6956) 
- [Support Backend plugins for Alert notifiers](https://github.com/grafana/grafana/issues/16004)
- [Custom Headers on HTTP webhook](https://github.com/grafana/grafana/issues/9813#top)
- [Support custom headers on the webhook contact point](https://github.com/grafana/grafana/issues/50558#top)
- [Webhook Receiver should support custom headers](https://github.com/grafana/alerting/issues/134#top)

Upstream Prometheus Alertmanager also has significant community requests, although consensus so far has been to keep the webhook notifier as-is with recommendations to set up a small web server to act as a webhook bridge:
- [it's more useful for webhook_config support template and custom the header](https://github.com/prometheus/alertmanager/issues/2268#top)
- [Added functionality to alter Webhook message](https://github.com/prometheus/alertmanager/pull/1422)
- [Reintroduce AWS SNS notifications](https://github.com/prometheus/alertmanager/pull/709#issuecomment-312595545)
- [Add templated text support for webhook receivers](https://github.com/prometheus/alertmanager/pull/2184#issuecomment-586647546)
- [Templating support for body of webhook_config](https://github.com/prometheus/alertmanager/issues/1496#issuecomment-409930740)
- [webhook custom style](https://github.com/prometheus/alertmanager/issues/3110#top)

## Future Work

- Import [gomplate](https://docs.gomplate.ca/) and expand template support to more namespaces.


## Custom Payload Examples

Could be as simple as writing JSON directly:
```json
{
  "alert_name": "{{ .CommonLabels.alertname }}",
  "status": "{{ .Status }}",
  "environment": "{{ .Vars.environment }}",
  "custom_field": "{{ .Vars.custom_field }}"
}
```
Or using new functions to help craft valid JSON:
```blade
  {{ coll.Dict 
  "receiver" .Receiver
  "status" .Status
  "alerts" .Alerts
  "groupLabels" .GroupLabels
  "commonLabels" .CommonLabels
  "commonAnnotations" .CommonAnnotations
  "externalURL" .ExternalURL
  "version" "1"
  "orgId"  (index .Alerts 0).OrgId
  "truncatedAlerts"  .TruncatedAlerts
  "groupKey" .GroupKey
  "state"  (tmpl.Inline "{{ if eq .Status \"resolved\" }}ok{{ else }}alerting{{ end }}" . )
  "title" (tmpl.Exec "default.title" . )
  "message" (tmpl.Exec "default.message" . )
  | data.ToJSONPretty " "}}
```

## Review Tips

- New gomplate-based functions are nearly 1:1 with the library, so testing was skipped given we're going to eventually remove these in favour of and import. Any differences are pointed out in comments.
- Grafana-side PR will be easier to test with as it will include UI support and enhancements to the template builder to help when creating JSON.

JacobsonMT committed:
Select gomplate template functions

JacobsonMT committed:
Custom headers

JacobsonMT committed:
Remove need for .Extra on ExtendedData by adding actual fields

JacobsonMT committed:
Support empty payload

JacobsonMT committed:
Webhook custom payload template

JacobsonMT committed:
Test fix

JacobsonMT committed:
Linting

moustafab reviewed (approved):
Looks great! Nice work!

My only thoughts on the implementation are around further improvements in the supported functions. If we anticipate significant churn here it may be good to establish an experimental namespace which we document to allow for a bit of safer additions. I'm hesitant to support to much without clear needs and I think this approach you've taken so far is far preferable to pulling in an entire library until clear need is well documented/established.

Also, templating header values might come in handy in the future but maybe we can wait for a request from the community before adding that.

titolins commented on receivers/webhook/webhook.go:
```diff
@@ -66,6 +66,14 @@ func (wn *Notifier) Notify(ctx context.Context, as ...*types.Alert) (bool, error
+	if tmplErr != nil {
+		return false, tmplErr
+	}
+	tmplErr = nil // Reset the error for the next template.
```
nit: this shouldn't be required, as cases where `tmplErr` is not nil will be caught by the if above. it might help make some sense of this implementation though 😅

titolins commented on templates/gomplate/data.go:
```diff
@@ -0,0 +1,87 @@
+package gomplate
```
mostly curious about such cases, but should we include gomplate's license?

JacobsonMT replied:
Good question, I'm not sure. I'll ask around.

JacobsonMT replied:
Added their MIT license to the individual gomplate files evolved from https://github.com/hairyhenderson/gomplate

titolins commented on receivers/webhook/config.go:
```diff
@@ -99,5 +121,61 @@ func NewConfig(jsonData json.RawMessage, decryptFn receivers.DecryptFunc) (Confi
+			omitted = append(omitted, k)
+			continue
+		}
+		safeHeaders[http.CanonicalHeaderKey(k)] = v
```
I do like the idea of standardizing the headers, but it could prove troublesome for users. Sometimes users might not have complete control over the receiver's implementation and that might make use of non-"canonicalized" headers.. I've had issues using [header.Add](https://pkg.go.dev/net/http#Header.Add) before because of that.
It still makes sense using it for comparison, but maybe it's worth keeping the safe headers in the original format. WDYT?

JacobsonMT replied:
:+1: This is a very good point and 100% correct IMO. I'll keep them as-is.

titolins reviewed (approved):
weekly board maintenance ✅ 😂 
great work 👏

JacobsonMT committed:
Address review comments

- Keep original case of headers
- Remove redundant tmplErr nil check but keep comment

JacobsonMT committed:
Add license info to individual files copied from gomplate
